### PR TITLE
[SPARK-32332][SQL] Support columnar exchanges when AQE is enabled

### PR DESCRIPTION
**Pull request by andygrove**
### What changes were proposed in this pull request?

This PR adds support for plugins to be able to provide columnar exchanges when AQE is enabled.

The main changes are:

- In AdaptiveSparkPlanExec, the newQueryStage method now optimizes the Exchange node rather than just optimizing its children. This allows plugins to participate in optimizing and replacing the Exchange node itself.
- New abstract base classes have been introduced for ShuffleExchange and BroadcastExchange so that Spark can work with replaced versions of these operators.

### Why are the changes needed?

The changes are needed so that plugins can provide columnar exchage operators when AQE is enabled.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Existing unit tests have been updated as part of this PR to test replacing broadcast and shuffle exchanges when AQE is enabled.

### CODE REVIEW

**Comment by tgravescs** [comment]
ok to test

File: sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala [hunk 1]
```diff
@@ -145,33 +150,56 @@ class SparkSessionExtensionSuite extends SparkFunSuite {
...
       // Verify that both pre and post processing of the plan worked.
-      val found = df.queryExecution.executedPlan.collect {
-        case rep: ReplacedRowToColumnarExec => 1
-        case proj: ColumnarProjectExec => 10
-        case c2r: ColumnarToRowExec => 100
-      }.sum
-      assert(found == 111)
-
+      val found = collectPlanSteps(df.queryExecution.executedPlan).sum
+      assert(found == 11121)
```
Comment (tgravescs) [inline sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala:197]: might be nice to comment what 11121 equals in terms of the execs - MyBroadcastExchangeExec, etc..

**Comment by tgravescs** [comment]
overall approach makes sense to me. It would be great to get some feedback from people working on AQE - @maryannxue @cloud-fan

**Comment by tgravescs** [comment]
org.apache.spark.sql.execution.adaptive.AdaptiveQueryExecSuite is failing @andygrove please take a look

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala [hunk 2]
```diff
@@ -37,16 +37,30 @@ import org.apache.spark.sql.execution._
...
 import org.apache.spark.util.collection.unsafe.sort.{PrefixComparators, RecordComparator}
 
+/**
+ * Base class for implementations of shuffle exchanges. This was added to enable plugins to
+ * provide columnar implementations of shuffle exchanges when Adaptive Query Execution is
+ * enabled.
+ */
+abstract class ShuffleExchange extends Exchange {
+  def shuffleId: Int
+  def getNumMappers: Int
```
Comment (maryannxue) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala:51]: We can get this through MapOutputTracker.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala [hunk 3]
```diff
@@ -37,16 +37,30 @@ import org.apache.spark.sql.execution._
...
 import org.apache.spark.util.MutablePair
 import org.apache.spark.util.collection.unsafe.sort.{PrefixComparators, RecordComparator}
 
+/**
+ * Base class for implementations of shuffle exchanges. This was added to enable plugins to
+ * provide columnar implementations of shuffle exchanges when Adaptive Query Execution is
+ * enabled.
+ */
+abstract class ShuffleExchange extends Exchange {
+  def shuffleId: Int
```
Comment (maryannxue) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala:50]: This is available in `mapOutputStats`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala [hunk 4]
```diff
@@ -374,12 +374,14 @@ case class AdaptiveSparkPlanExec(
   }
 
   private def newQueryStage(e: Exchange): QueryStageExec = {
-    val optimizedPlan = applyPhysicalRules(e.child, queryStageOptimizerRules)
+    // apply optimizer rules to the Exchange node and its children, allowing plugins to be
+    // able to replace the Exchange node itself
+    val optimizedPlan = applyPhysicalRules(e, queryStageOptimizerRules)
```
Comment (maryannxue) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:379]: How can we guarantee the top node is still an Exchange after applying physical rules?

Comment (maryannxue) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:379]: I think it's safer to use `s.withNewChildren(optimizedChildPlan)` than adding special handling in those physical rules: https://github.com/apache/spark/pull/29134/files#diff-a30c7a6fcdcdd13e57135fd04d05f3b7R115-R117

That saves you the trouble of worrying about certain assumptions being broken in an arbitrary rule.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala [hunk 5]
```diff
@@ -202,7 +202,7 @@ case class OptimizeSkewedJoin(conf: SQLConf) extends Rule[SparkPlan] {
         val leftParts = if (isLeftSkew && !isLeftCoalesced) {
           val reducerId = leftPartSpec.startReducerIndex
           val skewSpecs = createSkewPartitionSpecs(
-            left.shuffleStage.shuffle.shuffleDependency.shuffleId, reducerId, leftTargetSize)
```
Comment (maryannxue) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala:205]: Get shuffleId from mapStats.

**Comment by cloud-fan** [comment]
Hi @andygrove and @tgravescs , actually we have done something similar in our fork, to support columnar exchange. Our version supports broadcast as well, and also works with AQE disabled. Do you mind if I push our version out and ask you to review it? Thanks in advance!

**Comment by tgravescs** [comment]
@cloud-fan that sounds good, push it out and we'll take a look.

**Comment by tgravescs** [comment]
@cloud-fan do you know when you might have PR up?

**Comment by andygrove** [comment]
Replaced by https://github.com/apache/spark/pull/29262

**Pull request closed without merging**